### DISCUSSION TITLE
Build fails at final linking step

### DISCUSSION THREAD

**Discussion by ckcollab** [InteractiveComputerGraphics/SPlisHSPlasH#213] [Q&A]
Here's the Dockerfile I'm using, which uses CUDA 11.3 and the latest commit of PySplishSplash. This configuration gets the furthest of anything yet, but I did also try CUDA 11.8 and CUDA 10.2

```Dockerfile
FROM nvidia/cuda:11.3.1-cudnn8-devel-ubuntu20.04

# Avoid requiring input during installs
ENV DEBIAN_FRONTEND=noninteractive

# Pysplishsplash deps
RUN apt-get update && apt-get install -y git cmake xorg-dev freeglut3-dev build-essential

# Setup x server stuff to view the simulations
RUN apt-get install -y xvfb x11vnc
RUN mkdir ~/.vnc
RUN x11vnc -storepasswd secret ~/.vnc/passwd

ENV DISPLAY=":0"

# Do our noodling in temp dir
WORKDIR /tmp

# Install python stuff
RUN apt install -y python3 python3-setuptools python3-pip
ADD requirements.txt /requirements.txt
RUN pip install -r /requirements.txt

# Install pysplishsplash
RUN git clone https://github.com/InteractiveComputerGraphics/SPlisHSPlasH.git && \
    cd SPlisHSPlasH && \
    git checkout f46998f11c36e3735ac44560b49e2c85dcf3e9b4 && \
    python setup.py bdist_wheel -DCMAKE_BUILD_TYPE=Release -DUSE_AVX=On -DUSE_OpenMP=On -DUSE_IMGUI=On -DUSE_GPU_NEIGHBORHOOD_SEARCH=On && \
    pip install -I build/dist/*.whl && \
    rm -rf /tmp/SPlisHSPlasH

# Entrypoint
WORKDIR /app
ADD docker/startup.sh /startup.sh
ENTRYPOINT /startup.sh
```

Here's the error I get riiiiiiight as it's about to finish, such a bummer!

```bash
#0 309.5 [100%] Linking CXX shared module ../../lib.linux-x86_64-3.8/pysplishsplash.cpython-38-x86_64-linux-gnu.so
#0 341.8 /usr/bin/ld: ../extern/install/NeighborhoodSearch/lib/libcuNSearch.a(cuNSearch.cu.o): relocation R_X86_64_PC32 against symbol `_ZGVZN6thrust2mr19get_global_resourceINS_26device_ptr_memory_resourceINS_6system4cuda6detail20cuda_memory_resourceIXadL_Z10cudaMallocEEXadL_Z8cudaFreeEENS_8cuda_cub7pointerIvEE
EEEEEEPT_vE8resource' can not be used when making a shared object; recompile with -fPIC
#0 341.8 /usr/bin/ld: final link failed: bad value
#0 341.8 collect2: error: ld returned 1 exit status
#0 341.8 make[3]: *** [pySPlisHSPlasH/CMakeFiles/pysplishsplash.dir/build.make:524: ../lib.linux-x86_64-3.8/pysplishsplash.cpython-38-x86_64-linux-gnu.so] Error 1
#0 341.8 make[2]: *** [CMakeFiles/Makefile2:1428: pySPlisHSPlasH/CMakeFiles/pysplishsplash.dir/all] Error 2
#0 341.8 make[1]: *** [CMakeFiles/Makefile2:1435: pySPlisHSPlasH/CMakeFiles/pysplishsplash.dir/rule] Error 2
#0 341.8 make: *** [Makefile:482: pysplishsplash] Error 2
```

As best I can tell, this error doesn't make much sense, and I'm not sure where to try and insert `-fPIC` -- it seems like that argument may need to go somewhere relevant to `cuNSearch`?

Any help is much appreciated! Will keep digging.

**Comment by janbender** [comment]
So far we only tested the lib with Cuda 10.1. If you want to try with the "-fPIC" option you can add this in the file CMake/Common.cmake to CMAKE_CXX_FLAGS_RELEASE (around line 90). However, this will only influence SPlisHSPlasH, not cuNSearch. To modifiy cuNSearch, you have to take a look in <build-folder>/extern after the first build which downloads the code.

**Reply by ckcollab** [reply]
After trying multiple versions of GCC, Clang, setting `-fPIC` various ways.. I noticed this line:

https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/blob/f46998f11c36e3735ac44560b49e2c85dcf3e9b4/CMake/Common.cmake#L46

Should `-fPIC` already be being set?!

Here's what I'm calling, for reference:

```bash
$ python3.8 setup.py bdist_wheel -DCMAKE_BUILD_TYPE=Release -DUSE_AVX=On -DUSE_OpenMP=On -DUSE_IMGUI=On -DUSE_GPU_NEIGHBORHOOD_SEARCH=On -DCUSE_PYTHON_BINDINGS=On 
```

Also tried with `-DCMAKE_CXX_FLAGS=-fPIC` and such. Also, put `set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")` right before line 90 as suggested...

**Comment by janbender** [comment]
The question is if it is already set for cuNSearch which has own CMake files and build options. So I would suggest to add the option there.
